pkg/provisioner/crdprovisioner.go

```
@@ -375,72 +380,59 @@ func (c *CrdProvisioner) PublishVolume(
-			}
+		// if maxMountReplicaCount has been exceeded, unpublish demoted AzVolumeAttachment or if demoted AzVolumeAttachment does not exist, select one to unpublish
+		if len(existingAttachments) > azVolume.Spec.MaxMountReplicaCount {
+			roleChangeLabel, _ := azureutils.CreateLabelRequirements(consts.RoleChangeLabel, selection.Equals, consts.Demoted)
+			demotedAttachments, err := azVALister.List(azVASelector.Add(*roleChangeLabel))
 
-			// Otherwise, we are trying to update
-			// the updateInstanceolumeAttachment role from Replica to Primary
-			updateInstance.Spec.RequestedRole = diskv1alpha2.PrimaryRole
-			// Keeping the spec fields up to date with the request parameters
-			updateInstance.Spec.VolumeContext = volumeContext
-			// Update the label of the AzVolumeAttachment
-			if updateInstance.Labels == nil {
-				updateInstance.Labels = map[string]string{}
+			// if there is no demoted AzVolumeAttachment, randomly pick a replica to unpublish (TODO: come up with a logic to select replica for unpublishing)
+			var attachmentToUnpublish *diskv1alpha2.AzVolumeAttachment
+			if err != nil || len(demotedAttachments) == 0 {
+				attachmentToUnpublish = existingAttachments[0]
+			} else {
+				attachmentToUnpublish = demotedAttachments[0]
 			}
-			updateInstance.Labels[consts.RoleLabel] = string(diskv1alpha2.PrimaryRole)
 
-			return nil
+			if err := c.UnpublishVolume(ctx, attachmentToUnpublish.Spec.VolumeID, attachmentToUnpublish.Spec.NodeName, nil); err != nil {
+				return nil, status.Errorf(codes.Internal, "failed to make request to unpublish volume (%s) from node (%s): %v", attachmentToUnpublish.Spec.VolumeName, attachmentToUnpublish.Spec.NodeName, err)
+			}
 		}
+	}
+
+	attachmentName := azureutils.GetAzVolumeAttachmentName(volumeName, nodeID)
+
+	attachmentObj, err := azVALister.Get(attachmentName)
+
+	// if replica AzVolumeAttachment CRI does not exist for the volume-node pair, create one
+	if errors.IsNotFound(err) {
```

If `azVALister.Get(attachmentName)` failed with an error other than "not found" should we return here rather than assuming the object is good later?